pkg/lib/aws/ecr.go

```diff
@@ -0,0 +1,74 @@
+/*
+Copyright 2020 Cortex Labs, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package aws
+
+import (
+	"encoding/base64"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/service/ecr"
+	"github.com/cortexlabs/cortex/pkg/lib/errors"
+	"github.com/cortexlabs/cortex/pkg/lib/regex"
+)
+
+type ECRAuthConfig struct {
+	Username      string
+	AccessToken   string
+	ProxyEndpoint string
+}
+
+func (c *Client) GetECRAuthToken() (*ecr.GetAuthorizationTokenOutput, error) {
+	svc := c.ECR()
+	result, err := svc.GetAuthorizationToken(&ecr.GetAuthorizationTokenInput{})
+	if err != nil {
+		return result, errors.Wrap(err, "failed to retrieve ECR auth token")
+	}
+	return result, nil
+}
+
+func (c *Client) GetECRAuthConfig() (ECRAuthConfig, error) {
+	tokenOutput, err := c.GetECRAuthToken()
+	if err != nil {
+		return ECRAuthConfig{}, err
+	}
+	authData := tokenOutput.AuthorizationData[0]
```

Comment:
Is `len(tokenOutput.AuthorizationData) > 0` guaranteed?

Reply:
Judging from what I've seen in the AWS SDK docs, yes, that appears to be guaranteed unless you get an error. Apparently, there has to be a 1:1 mapping between what's being requested and what you get back, otherwise, you get a non-200 response. At least that's my understanding. It doesn't help that the docs are not overly clear about that. Here are the docs: https://docs.aws.amazon.com/AmazonECR/latest/APIReference/API_GetAuthorizationToken.html

What do you think of checking the length and if not return an error?

Reply:
Yeah, we have been bit by something like this with the AWS SDK before, so I think we might as well do the check if it's straightforward

Reply:
I did a `return ..., ErrorECRExtractingCredentials()` if `AuthorizationData` is empty. Do you think that's not enough, or should the error's message be more clear?

Reply:
I think that is sufficient, given we do not expect this to happen. Mostly I was motivated to avoid the nil pointer.